presto-main/src/main/java/com/facebook/presto/metadata/PolymorphicScalarFunctionBuilder.java

```diff
@@ -243,33 +217,117 @@ public MethodsGroupBuilder withExtraParameters(Function<SpecializeContext, List<
     }
 
-    static class MethodsGroup
+    public static class ChoiceBuilder
+    {
+        private final Class<?> clazz;
+        private final Signature signature;
+        private boolean nullableResult;
+        private List<ArgumentProperty> argumentProperties;
+        private final ImmutableList.Builder<MethodsGroup> methodsGroups = ImmutableList.builder();
+
+        private ChoiceBuilder(Class<?> clazz, Signature signature)
+        {
+            this.clazz = requireNonNull(clazz, "clazz is null");
+            this.signature = requireNonNull(signature, "signature is null");
+        }
+
+        public ChoiceBuilder implementation(Function<MethodsGroupBuilder, MethodsGroupBuilder> methodsGroupSpecification)
+        {
+            MethodsGroupBuilder methodsGroupBuilder = new MethodsGroupBuilder(clazz);
+            methodsGroupSpecification.apply(methodsGroupBuilder);
+            methodsGroups.add(methodsGroupBuilder.build());
+            return this;
+        }
+
+        public ChoiceBuilder nullableResult(boolean nullableResult)
+        {
+            this.nullableResult = nullableResult;
+            return this;
+        }
+
+        public ChoiceBuilder argumentProperties(ArgumentProperty... argumentProperties)
+        {
+            requireNonNull(argumentProperties, "argumentProperties is null");
+            this.argumentProperties = ImmutableList.copyOf(argumentProperties);
+            return this;
+        }
+
+        public PolymorphicScalarFunctionChoice build()
+        {
+            if (argumentProperties == null) {
+                argumentProperties = nCopies(signature.getArgumentTypes().size(), valueTypeArgumentProperty(RETURN_NULL_ON_NULL));
+            }
+            return new PolymorphicScalarFunctionChoice(nullableResult, argumentProperties, methodsGroups.build());
+        }
+    }
+
+    static final class MethodsGroup
     {
         private final List<Method> methods;
         private final Optional<Function<SpecializeContext, List<Object>>> extraParametersFunction;
+        private final List<NullConventionAwareMethod> nullConventionAwareMethods;
 
-        private MethodsGroup(
+        MethodsGroup(
                 List<Method> methods,
-                Optional<Function<SpecializeContext, List<Object>>> extraParametersFunction)
+                Optional<Function<SpecializeContext, List<Object>>> extraParametersFunction,
+                List<NullConventionAwareMethod> nullConventionAwareMethods)
         {
             this.methods = requireNonNull(methods, "methods is null");
             this.extraParametersFunction = requireNonNull(extraParametersFunction, "extraParametersFunction is null");
+            this.nullConventionAwareMethods = requireNonNull(nullConventionAwareMethods, "methodBlockPositionTypes is null");
         }
 
         List<Method> getMethods()
         {
             return methods;
         }
 
+        List<Optional<Class<?>>> findBlockPositionTypes(Method method)
```

Comment:
Make `getMethods` method return a list of `MethodAndNativeContainerTypes`. This way, you can remove the `findBlockPositionTypes` method.

Reply:
I removed this function since getMethods return a list of MethodAndNativeContainerTypes